runtime/vm/jnifield.cpp

```diff
@@ -33,407 +33,886 @@
+	jdouble value = *((jdouble *)&uvalue);
 
 	if (j9FieldID->field->modifiers & J9AccVolatile) {
-		VM_AtomicSupport::readWriteBarrier();
+		VM_AtomicSupport::readBarrier();
 	}
 
-	FieldEvents::triggerSetEvents(vmThread, j9FieldID, object, &value);
-	VM_VMAccess::inlineExitVMToJNI(vmThread);
+	VM_VMAccess::inlineExitVMToJNI(currentThread);
+	return value;
 }
 
 void JNICALL
 setByteField(JNIEnv *env, jobject obj, jfieldID fieldID, jbyte value)
 {
-	putPrimitiveField<ByteField, jbyte>(env, obj, fieldID, value);
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	VM_VMAccess::inlineEnterVMFromJNI(currentThread);
+	J9JNIFieldID *j9FieldID = (J9JNIFieldID *)fieldID;
+	UDATA valueOffset = j9FieldID->offset;
+
+	if (J9_EVENT_IS_HOOKED(vm->hookInterface, J9HOOK_VM_PUT_FIELD)) {
+		j9object_t object = J9_JNI_UNWRAP_REFERENCE(obj);
+		if (J9_ARE_ANY_BITS_SET(J9OBJECT_CLAZZ(currentThread, object)->classFlags, J9ClassHasWatchedFields)) {
+			IDATA location = 0;
+			J9Method *method = findFieldContext(currentThread, &location);
+			if (NULL != method) {
+				U_64 newValue = 0;
+				*(I_32*)&newValue = (I_32)value;
+				ALWAYS_TRIGGER_J9HOOK_VM_PUT_FIELD(vm->hookInterface, currentThread, method, location, object, j9FieldID->offset, newValue);
+			}
+		}
+	}
+
+	bool isVolatile = J9_ARE_ANY_BITS_SET(j9FieldID->field->modifiers, J9AccVolatile);
+	if (isVolatile) {
+		VM_AtomicSupport::writeBarrier();
+	}
+
+	j9object_t object = J9_JNI_UNWRAP_REFERENCE(obj);
+	valueOffset += J9_OBJECT_HEADER_SIZE;
+	J9OBJECT_U32_STORE(currentThread, object, valueOffset, (U_32)value);
+
+	if (isVolatile) {
+		VM_AtomicSupport::readWriteBarrier();
+	}
+
+	VM_VMAccess::inlineExitVMToJNI(currentThread);
 }
 
 void JNICALL
 setBooleanField(JNIEnv *env, jobject obj, jfieldID fieldID, jboolean value)
 {
-	putPrimitiveField<BooleanField, jboolean>(env, obj, fieldID, value);
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	VM_VMAccess::inlineEnterVMFromJNI(currentThread);
+	J9JNIFieldID *j9FieldID = (J9JNIFieldID *)fieldID;
+	UDATA valueOffset = j9FieldID->offset;
+
+	if (J9_EVENT_IS_HOOKED(vm->hookInterface, J9HOOK_VM_PUT_FIELD)) {
+		j9object_t object = J9_JNI_UNWRAP_REFERENCE(obj);
+		if (J9_ARE_ANY_BITS_SET(J9OBJECT_CLAZZ(currentThread, object)->classFlags, J9ClassHasWatchedFields)) {
+			IDATA location = 0;
+			J9Method *method = findFieldContext(currentThread, &location);
+			if (NULL != method) {
+				U_64 newValue = 0;
+				*(I_32*)&newValue = (I_32)value;
+				ALWAYS_TRIGGER_J9HOOK_VM_PUT_FIELD(vm->hookInterface, currentThread, method, location, object, j9FieldID->offset, newValue);
+			}
+		}
+	}
+
+	bool isVolatile = J9_ARE_ANY_BITS_SET(j9FieldID->field->modifiers, J9AccVolatile);
+	if (isVolatile) {
+		VM_AtomicSupport::writeBarrier();
+	}
+
+	j9object_t object = J9_JNI_UNWRAP_REFERENCE(obj);
+	valueOffset += J9_OBJECT_HEADER_SIZE;
```

Comment:
Does this need to do an `& 1` when storing?

Reply:
I think the answer is yes, but it never has done.  Can we leave this to a future PR once we decide how we're going to enforce the type consistency throughout the VM?

Reply:
Fair enough.